pixeltable/utils/parquet.py

```diff
@@ -0,0 +1,132 @@
+
+
+def save_parquet(df : DataFrame, dest_path : Path, partition_size_bytes : int = 100_000_000):
+    """
+        Internal method to stream dataframe data to parquet format.
+        Does not materialize the dataset to memory.
+
+        It preserves pixeltable type metadata in a json file, which would otherwise 
+        not be available in the parquet format.
+
+        Images are stored inline in a compressed format in their parquet file.
+
+        NB. pixeltable types
+            - string ok
+            - image ok (inlined)
+            - video ok (path only)
+    
+        Args:
+            partition_size_bytes: maximum target size for each chunk. Default 100MB.
+        Returns:
+            path to the parquet directory
+
+    """
+    column_names = df.get_column_names()
+    column_types = df.get_column_types()
+
+    arrow_schema = pa.schema([pa.field(name, column_types[i].to_arrow_type()) for i, name in enumerate(column_names)])
+    # store the changes atomically
+    with transactional_folder(dest_path) as temp_path:
+        # dump metadata json file so we can inspect what was the source of the parquet file later on.
+        json.dump(df.as_dict(), (temp_path / '.pixeltable.json').open('w')) # pylint: disable=protected-access
+
+        batch_num = 0
+        current_value_batch : Dict[str, deque] = {k:deque() for k in column_names}
+        current_byte_estimate = 0
+
+        for data_row in df._exec(): # pylint: disable=protected-access
+            for (col_name, col_type, e) in zip(column_names, column_types, df._select_list_exprs): # pylint: disable=protected-access
+                val = data_row[e.slot_idx]
+                length = None
+                if col_type.is_image_type():
+                    # images get inlined into the parquet file
+                    if data_row.file_paths is not None and data_row.file_paths[e.slot_idx] is not None: 
+                        # if there is a file, read directly to preserve information
+                        val = open(data_row.file_paths[e.slot_idx], 'rb').read()
+                    elif isinstance(val, PIL.Image.Image): 
+                        # if no file available, eg. bc it is computed, convert to png
+                        buf = io.BytesIO()
+                        val.save(buf, format='PNG')
+                        val = buf.getvalue()
+                    else:
+                        assert False, f'unknown image type {type(val)}'
+                    length = len(val)
+                elif col_type.is_string_type():
+                    length = len(val)
+                elif col_type.is_video_type():
+                    if data_row.file_paths is not None and data_row.file_paths[e.slot_idx] is not None: 
+                        # if there is a file, read directly to preserve information
+                        val = data_row.file_paths[e.slot_idx]
+                    else:
+                        assert False, f'unknown video type {type(val)}'
```

Comment:
confusing error message, the type doesn't really matter here.

does it even make sense to include file names directly? should 'video' not be a supported type for the purpose of pytorch conversion?